src/Shared/Microsoft.DotNet.Web.Authentication.Tests/GitHubClaimsResolverTests.cs

```diff
@@ -0,0 +1,508 @@
+        {
+            _output = output;
+        }
+
+        [Fact]
+        public async Task GitHubErrorFailedAuth()
+        {
+            using IDisposable scope = ConfigureResolver(out FakeHandler handler, out _, out GitHubClaimResolver resolver);
+
+            handler.AddCannedResponse(HttpStatusCode.TooManyRequests,
+                "https://api.github.test/user",
+                new JObject
+                {
+                    {"error", "Rate limit exceeded"},
+                    {"really-long-details", new string('*', 10000)},
+                });
+
+            await Assert.ThrowsAsync<HttpRequestException>(() => resolver.GetUserInformationClaims("FAKE-TOKEN"));
+
+            handler.AssertCompleted();
+        }
+
+        [Fact]
+        public async Task UserInformationIsPopulated()
+        {
+            using IDisposable scope =
+                ConfigureResolver(out FakeHandler handler, out _, out GitHubClaimResolver resolver);
+
+            handler.AddCannedResponse("https://api.github.test/user",
+                new JObject
+                {
+                    {"id", 146},
+                    {"login", "TestUser"},
+                    {"email", "TestEmail@microsoft.test"},
+                    {"name", "A Real Fake Name"},
+                    {"url", "https://github.test/TestUser"},
+                });
+
+            (IEnumerable<Claim> claims, JObject job) = await resolver.GetUserInformation("FAKE-TOKEN");
+            Assert.NotNull(job);
+            var id = new ClaimsIdentity(claims, "TEST");
+            var principal = new ClaimsPrincipal(id);
+            string accessToken = resolver.GetAccessToken(principal);
+
+            Assert.Equal("FAKE-TOKEN", accessToken);
+            Assert.Equal("TestUser", id.Name);
+            Assert.Equal("TestEmail@microsoft.test", id.Claims.FirstOrDefault(c => c.Type == ClaimTypes.Email)?.Value);
+
+            handler.AssertCompleted();
+        }
+
+        [Fact]
+        public async Task UserInformationIsCached()
+        {
+            using IDisposable scope =
+                ConfigureResolver(out FakeHandler handler, out _, out GitHubClaimResolver resolver);
+
+            handler.AddCannedResponse("https://api.github.test/user",
+                new JObject
+                {
+                    {"id", 146},
+                    {"login", "TestUser"},
+                    {"email", "TestEmail@microsoft.test"},
+                    {"name", "A Real Fake Name"},
+                    {"url", "https://github.test/TestUser"},
+                });
+
+            IEnumerable<Claim> userInformation = await resolver.GetUserInformationClaims("FAKE-TOKEN");
+            var id = new ClaimsIdentity(userInformation, "TEST");
+            Assert.Equal("TestUser", id.Name);
+
+            userInformation = await resolver.GetUserInformationClaims("FAKE-TOKEN");
+            id = new ClaimsIdentity(userInformation, "TEST");
```

Comment:
I'm assuming that this works because the "AddCannedResponse" only responds once. If this second call makes another http request this will fail. I would call this out in the test if that is correct.

Reply:
Yup. I'll add a comment. :-)